Fixes #15739: Compact REPL printouts of an `Associative` use brackets when appropriate

Tetralux opened the pull request:
Displaying a `Dict` in compact format manually prints out the pairs in the form `key=>value` which is problematic if either the key or value is a `Pair`, itself.

This PR defers the printing of the ~~keys and the values~~ key-value pairs to `show`.

E.g:

On master:

``` julia
julia> Dict( (1=>2)=>(3=>10), (2=>4)=>(6=>20) )
Dict(2=>4=>6=>20,1=>2=>3=>10)
```

With this PR: 

``` julia
julia> Dict( (1=>2)=>(3=>10), (2=>4)=>(6=>20) )
Dict((2=>4)=>(6=>20),(1=>2)=>(3=>10))
```

Tetralux committed:
`Pair`s of `Pair`s should enclose their parts in brackets. Fixes #15739

Tetralux committed:
Add test

tkelman commented:
Thanks! Looks like the display order differs on 32 bit (different hashing results) which is causing the tests to fail on those platforms.

Tetralux committed:
Fix test not working on all CPU types

Dictionary display order is different depending on bitness of the executing CPU.

H-225 commented:
@tkelman Oops! And here's take two. Or at least, when the Travis queue gets to it.

Keno commented on test/dict.jl:
```diff
@@ -294,6 +294,18 @@ for d in (Dict("\n" => "\n", "1" => "\n", "\n" => "2"),
+    # each time.
+    expectedstr = join([string(x) for x in d], ",")
+
+    @test bytestring(buf) == "Dict($expectedstr)"
```
Why not check whether it contains both expected strings. If somebody were to say break printing of Pairs, this would still pass.

H-225 replied:
@Keno A fair point. I'll have it check for all the strings we care about.

Keno commented:
LGTM, please squash the commits if you don't mind.

Keno commented:
Actually, let's try github's newfangled squash/merge. Hold on.

tkelman commented:
It annoyingly drops the commit messages for later commits, and I kind of miss the "merge PR" reference in the log but it's nice that github at least gives the option now.

H-225 commented:
@tkelman @Keno Thank you! :smile:

H-225 commented:
@Keno Is it wise to rebase this branch and/or reuse it to fix the problem at https://github.com/JuliaLang/julia/issues/15739 - or shall I make a new PR?

Keno commented:
Please make a new PR.

vtjnash commented:
to expand on what Keno said, you might be able to use the same branch name for a new PR, but github doesn't allow reopening a PR once it has been merged

H-225 commented:
@vtjnash Good to know. Thanks.
